File: mobile/android/app/src/main/java/tr/gov/ade/features/integrations/IntegrationsViewModel.kt
package tr.gov.ade.features.integrations

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import tr.gov.ade.core.network.APIClient
import tr.gov.ade.data.model.*
import javax.inject.Inject

/**
 * Integrations ViewModel for ADE Android
 *
 * Handles integration data fetching, connection, and sync
 *
 * @author ADE Mobile Team
 * @since 2026-01-24
 */

@HiltViewModel
class IntegrationsViewModel @Inject constructor(
    private val apiClient: APIClient
) : ViewModel() {

    // MARK: - State

    private val _integrations = MutableStateFlow<List<Integration>>(emptyList())
    val integrations: StateFlow<List<Integration>> = _integrations.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    // MARK: - Data Loading

    fun loadIntegrations() {
        viewModelScope.launch {
            _isLoading.value = true

            apiClient.safeApiCall {
                apiClient.service.getIntegrations()
            }.onSuccess { response ->
                _integrations.value = response
            }.onFailure { error ->
                _error.value = error.message
            }

            _isLoading.value = false
        }
    }

    // MARK: - Actions

    fun connect(integration: Integration, credentials: Map<String, String>) {
        viewModelScope.launch {
            _isLoading.value = true

            val request = ConnectIntegrationRequest(
                credentials = credentials,
                settings = null
            )

            apiClient.safeApiCall {
                apiClient.service.connectIntegration(integration.id, request)
            }.onSuccess { response ->
                // Update local integration
                val updatedList = _integrations.value.map {
                    if (it.id == integration.id) response.integration else it
                }
                _integrations.value = updatedList

                println("✅ Connected to ${integration.name}")
            }.onFailure { error ->
                _error.value = error.message
            }

            _isLoading.value = false
        }
    }

    fun disconnect(integration: Integration) {
        viewModelScope.launch {
            _isLoading.value = true

            apiClient.safeApiCall {
                apiClient.service.disconnectIntegration(integration.id)
            }.onSuccess {
                // Update local integration
                val updatedList = _integrations.value.map {
                    if (it.id == integration.id) {
                        it.copy(isConnected = false, connectedAt = null)
                    } else {
                        it
                    }
                }
                _integrations.value = updatedList

                println("❌ Disconnected from ${integration.name}")
            }.onFailure { error ->
                _error.value = error.message
            }

            _isLoading.value = false
        }
    }

    fun syncAll() {
        viewModelScope.launch {
            _isLoading.value = true

            // Sync all connected integrations
            _integrations.value.filter { it.isConnected }.forEach { integration ->
                sync(integration)
            }

            _isLoading.value = false
        }
    }

    private suspend fun sync(integration: Integration) {
        apiClient.safeApiCall {
            apiClient.service.getIntegrationDetail(integration.id)
        }.onSuccess {
            println("🔄 Synced ${integration.name}")
        }.onFailure { error ->
            println("⚠️ Failed to sync ${integration.name}: ${error.message}")
        }
    }

    fun clearError() {
        _error.value = null
    }
}

File: mobile/ios/ADE/Features/Integrations/IntegrationsViewModel.swift
import Foundation
import Combine

/**
 * Integrations ViewModel for ADE iOS
 *
 * Handles integration data fetching, connection, and sync
 *
 * @author ADE Mobile Team
 * @since 2026-01-24
 */

@MainActor
class IntegrationsViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var integrations: [Integration] = []
    @Published var isLoading = false
    @Published var showError = false
    @Published var errorMessage: String?

    // MARK: - Dependencies

    private let apiClient = APIClient.shared
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Data Loading

    func loadIntegrations() async {
        isLoading = true

        do {
            let response: [Integration] = try await apiClient.request(
                endpoint: .getIntegrations
            )

            integrations = response

        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func filteredIntegrations(for category: IntegrationCategory) -> [Integration] {
        integrations.filter { $0.category == category }
    }

    // MARK: - Actions

    func connect(_ integration: Integration, credentials: [String: String]) async {
        isLoading = true

        do {
            let request = ConnectIntegrationRequest(
                credentials: credentials,
                settings: nil
            )

            let response: IntegrationDetail = try await apiClient.request(
                endpoint: .connectIntegration(id: integration.id),
                method: .post,
                body: request
            )

            // Update local integration
            if let index = integrations.firstIndex(where: { $0.id == integration.id }) {
                integrations[index] = response.integration
            }

            // Optional: Show success message
            print("✅ Connected to \(integration.name)")

        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func disconnect(_ integration: Integration) async {
        isLoading = true

        do {
            let _: EmptyResponse = try await apiClient.request(
                endpoint: .disconnectIntegration(id: integration.id),
                method: .post
            )

            // Update local integration
            if let index = integrations.firstIndex(where: { $0.id == integration.id }) {
                var updated = integrations[index]
                updated.isConnected = false
                updated.connectedAt = nil
                integrations[index] = updated
            }

            print("❌ Disconnected from \(integration.name)")

        } catch {
            errorMessage = error.localizedDescription
            showError = true
        }

        isLoading = false
    }

    func syncAll() async {
        isLoading = true

        // Sync all connected integrations
        for integration in integrations.filter({ $0.isConnected }) {
            await sync(integration)
        }

        isLoading = false
    }

    func sync(_ integration: Integration) async {
        do {
            let _: EmptyResponse = try await apiClient.request(
                endpoint: .getIntegrationDetail(id: integration.id),
                method: .post
            )

            print("🔄 Synced \(integration.name)")

        } catch {
            print("⚠️ Failed to sync \(integration.name): \(error.localizedDescription)")
        }
    }
}
